include/verilated.cpp

```diff
@@ -2790,26 +2791,36 @@ static struct {
     VoidPCbList s_exitCbs VL_GUARDED_BY(s_exitMutex);
 } VlCbStatic;
 
-static void addCb(Verilated::VoidPCb cb, void* datap, VoidPCbList& cbs) VL_MT_UNSAFE {
+static void addCbFlush(Verilated::VoidPCb cb, void* datap) VL_REQUIRES(VlCbStatic.s_flushMutex) {
     std::pair<Verilated::VoidPCb, void*> pair(cb, datap);
-    cbs.remove(pair);  // Just in case it's a duplicate
-    cbs.push_back(pair);
+    VlCbStatic.s_flushCbs.remove(pair);  // Just in case it's a duplicate
+    VlCbStatic.s_flushCbs.push_back(pair);
 }
-static void removeCb(Verilated::VoidPCb cb, void* datap, VoidPCbList& cbs) VL_MT_UNSAFE {
+static void addCbExit(Verilated::VoidPCb cb, void* datap) VL_REQUIRES(VlCbStatic.s_exitMutex) {
     std::pair<Verilated::VoidPCb, void*> pair(cb, datap);
-    cbs.remove(pair);
+    VlCbStatic.s_exitCbs.remove(pair);  // Just in case it's a duplicate
+    VlCbStatic.s_exitCbs.push_back(pair);
+}
+static void removeCbFlush(Verilated::VoidPCb cb, void* datap)
+    VL_REQUIRES(VlCbStatic.s_flushMutex) {
+    std::pair<Verilated::VoidPCb, void*> pair(cb, datap);
+    VlCbStatic.s_flushCbs.remove(pair);
+}
+static void removeCbExit(Verilated::VoidPCb cb, void* datap) VL_REQUIRES(VlCbStatic.s_exitMutex) {
+    std::pair<Verilated::VoidPCb, void*> pair(cb, datap);
+    VlCbStatic.s_exitCbs.remove(pair);
 }
 static void runCallbacks(const VoidPCbList& cbs) VL_MT_SAFE {
     for (const auto& i : cbs) i.first(i.second);
 }
 
 void Verilated::addFlushCb(VoidPCb cb, void* datap) VL_MT_SAFE {
     const VerilatedLockGuard lock{VlCbStatic.s_flushMutex};
```

Comment:
I think these should be VL_MT_ACQUIRES(VlCbStatic.s_flushMutex) or similar.

Reply:
You are right, we should mark it as ``VL_MT_SAFE_EXCLUDES(VlCbStatic.s_flushMutex)``, but I can't do this in header file as this struct is only defined in cpp file.

Instead I refactored the code, moved locking to to static functions and marked static functions as ``VL_MT_SAFE_EXCLUDES(VlCbStatic.s_flushMutex)``.